File: 2021_3/day9-Add One Row to Tree/python3/Add One Row to Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def addOneRow(self, root: TreeNode, v: int, d: int) -> TreeNode:
             
        if d==1:
            new_node = TreeNode(v)
            new_node.left = root
            return new_node
        
        elif d==2:
            nodeL = root.left
            nodeR = root.right
            
            root.left = TreeNode(v)
            root.left.left = nodeL
            
            root.right = TreeNode(v)
            root.right.right = nodeR
            
            return root
        
        if root.left: 
            # function name前要記得加self.
            self.addOneRow(root.left, v, d-1)
        if root.right:
            self.addOneRow(root.right, v, d-1)
        return root
            

File: 2021_3/day9-Add One Row to Tree/typescript/Add One Row to Tree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

 function addOneRow(root: TreeNode | null, v: number, d: number): TreeNode | null {
    if(!root)
        return root;
    
    if(d==1) {
        let new_node: TreeNode = new TreeNode(v);
        new_node.left = root;
        return new_node;
        
    } else if(d==2) {
        let nodeL: TreeNode = root.left;
        let nodeR: TreeNode = root.right;
        
        root.left = new TreeNode(v);
        root.left.left = nodeL;
        
        root.right = new TreeNode(v);
        root.right.right = nodeR;
        
        return root;
    }
    addOneRow(root.left, v, d-1);
    addOneRow(root.right, v, d-1);
    
    return root;
};